developer-workflow/extension-modules.rst

```diff
@@ -5,13 +5,679 @@
+Some modules in the standard library, such as :mod:`datetime` or :mod:`pickle`,
+have identical implementations in C and Python; the C implementation, when
+available, is expected to improve performance (such extension modules are
+commonly referred to as *accelerator modules*).
 
-Read the following references:
+Other modules mainly implemented in Python may import a C helper extension
+providing implementation details (for instance, the :mod:`csv` module uses
+the internal :mod:`!_csv` module defined in :cpy-file:`Modules/_csv.c`).
+
+For writing an extension module in C, we prefer to give you some links
+where you can read good documentation:
 
 * https://docs.python.org/dev/c-api/
 * https://docs.python.org/dev/extending/
 * :pep:`399`
 * https://pythonextensionpatterns.readthedocs.io/en/latest/
+
+Classifying extension modules
+=============================
+
+Extension modules can be classified into two categories:
+
+* A *built-in* extension module is a module built and shipped with
+  the Python interpreter. A built-in module is *statically* linked
+  into the interpreter, thereby lacking a :attr:`__file__` attribute.
+
+  .. seealso:: :data:`sys.builtin_module_names` --- names of built-in modules.
+
+  Built-in modules are built with the :c:macro:`!Py_BUILD_CORE_BUILTIN`
+  macro defined.
+
+* A *shared* (or *dynamic*) extension module is built as a shared library
+  (``.so`` or ``.dll`` file) and is *dynamically* linked into the interpreter.
+
+  In particular, the module's :attr:`__file__` attribute contains the path
+  to the ``.so`` or ``.dll`` file.
+
+  Shared modules are built with the :c:macro:`!Py_BUILD_CORE_MODULE`
+  macro defined. Using the :c:macro:`!Py_BUILD_CORE_BUILTIN` macro
+  instead causes an :exc:`ImportError` when importing the module.
+
+.. note::
+
+   Informally, built-in extension modules can be regarded as *required*
+   while shared extension modules are *optional* in the sense that they
+   might be supplied, overridden or disabled externally.
+
+   Usually, accelerator modules are built as *shared* extension modules,
+   especially if they already have a pure Python implementation.
+
+According to :pep:`399`, *new* extension modules MUST provide a working and
+tested pure Python implementation, unless a special dispensation is given.
+Please ask the :github:`Steering Council <python/steering-council>` if such
+dispensation is needed.
+
+Adding an extension module to CPython
+=====================================
+
+Assume that the standard library contains a pure Python module :mod:`!foo`
+together with the following :func:`!foo.greet` function:
+
+.. code-block:: python
+   :caption: Lib/foo.py
+
+   def greet():
+       return "Hello World!"
+
+Instead of using the Python implementation of :func:`!foo.greet`, we want to
+use its corresponding C implementation exposed in some :mod:`!_foo` module
+written in C. Ideally, we want to modify :cpy-file:`!Lib/foo.py` as follows:
```

Comment:
```suggestion
Instead of using the Python implementation of :func:`!foo.greet`, we want to
use its corresponding C extension implementation exposed in :mod:`!_foo` module.
Ideally, we want to modify :cpy-file:`!Lib/foo.py` as follows:
```